google/cloud/storage/internal/range_from_pagination.h

```diff
@@ -0,0 +1,180 @@
+#if GOOGLE_CLOUD_CPP_HAVE_CONST_REF_REF
+  value_type const&& operator*() const&& { return std::move(value_); }
+#endif  // GOOGLE_CLOUD_CPP_HAVE_CONST_REF_REF
+  value_type&& operator*() && { return std::move(value_); }
+
+  friend bool operator==(PaginationIterator const& lhs,
+                         PaginationIterator const& rhs) {
+    // All end iterators are equal.
+    if (lhs.owner_ == nullptr) {
+      return rhs.owner_ == nullptr;
+    }
+    // Iterators on different streams are always different.
+    if (lhs.owner_ != rhs.owner_) {
+      return false;
+    }
+    // Iterators on the same stream are equal if they point to the same object.
+    if (lhs.value_.ok() && rhs.value_.ok()) {
+      return lhs.value_.value() == rhs.value_.value();
+    }
+    return lhs.value_.status() == rhs.value_.status();
+  }
+
+  friend bool operator!=(PaginationIterator const& lhs,
+                         PaginationIterator const& rhs) {
+    return !(lhs == rhs);
+  }
+
+ private:
+  friend Range;
+  explicit PaginationIterator(Range* owner, value_type value)
+      : owner_(owner), value_(std::move(value)) {}
+
+ private:
+  Range* owner_;
+  value_type value_;
+};
+
+template <typename T, typename Request, typename Response>
+class PaginationRange {
+ public:
+  explicit PaginationRange(
+      Request request,
+      std::function<StatusOr<Response>(Request const& r)> loader)
+      : request_(std::move(request)),
+        next_page_loader_(std::move(loader)),
+        next_page_token_(),
+        on_last_page_(false) {
+    current_ = current_page_.begin();
+  }
+
+  /// The iterator type for this Range.
+  using iterator = PaginationIterator<T, PaginationRange>;
+
+  /**
+   * Return an iterator over the list of hmacKeys.
+   *
+   * The returned iterator is a single-pass input iterator that reads object
+   * metadata from the ListHmacKeysReader when incremented.
+   *
+   * Creating, and particularly incrementing, multiple iterators on the same
+   * ListHmacKeysReader is unsupported and can produce incorrect results.
+   */
+  iterator begin() { return GetNext(); }
+
+  /// Return an iterator pointing to the end of the stream.
+  iterator end() { return PaginationIterator<T, PaginationRange>{}; }
+
+ protected:
+  friend class PaginationIterator<T, PaginationRange>;
+
+  /**
+   * Fetches (or returns if already fetched) the next object from the stream.
+   *
+   * @return An iterator pointing to the next element in the stream. On error,
+   *   it returns an iterator that is different from `.end()`, but has an error
+   *   status. If the stream is exhausted, it returns the `.end()` iterator.
+   */
+  iterator GetNext() {
+    static Status const past_the_end_error(
+        StatusCode::kFailedPrecondition,
+        "Cannot iterating past the end of ListObjectReader");
+    if (current_page_.end() == current_) {
+      if (on_last_page_) {
+        return iterator(nullptr, StatusOr<T>(past_the_end_error));
```

Comment:
Why the difference between here and line 159?

Reply:
Why indeed?  There was not good reason, fixed.